sycl/source/detail/kernel_name_based_cache_t.hpp

```diff
@@ -0,0 +1,42 @@
+//==-------------------- kernel_name_based_cache_t.hpp ---------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#pragma once
+
+#include <detail/kernel_arg_mask.hpp>
+#include <sycl/detail/spinlock.hpp>
+#include <sycl/detail/ur.hpp>
+
+#include <mutex>
+#include <utility>
+
+#include <boost/unordered/unordered_flat_map.hpp>
+
+namespace sycl {
+inline namespace _V1 {
+namespace detail {
+using FastKernelCacheKeyT = std::pair<ur_device_handle_t, ur_context_handle_t>;
+using FastKernelCacheValT =
+    std::tuple<ur_kernel_handle_t, std::mutex *, const KernelArgMask *,
+               ur_program_handle_t>;
+using FastKernelSubcacheMapT =
+    ::boost::unordered_flat_map<FastKernelCacheKeyT, FastKernelCacheValT>;
+
+using KernelFastSubcacheMutexT = SpinLock;
+using KernelFastSubcacheReadLockT = std::lock_guard<KernelFastSubcacheMutexT>;
+using KernelFastSubcacheWriteLockT = std::lock_guard<KernelFastSubcacheMutexT>;
+
+using FastKernelSubcacheT =
+    std::pair<FastKernelSubcacheMapT, KernelFastSubcacheMutexT>;
```

Comment:
I strongly prefer a struct or class instead of just an `std::pair<>`. It is hard to understand the code when `std::pair` is used.
And for now, you can store the mutex just in the `struct KernelNameBasedCacheT` below. So you do not need to introduce an additional struct.

Reply:
Sure, I agree that would be better. I'd rather introduce a struct though and work with that in `KernelProgramCache` instead of exposing the entire `KernelNameBasedCacheT` to it. We would also need to handle creation/destruction in the wrapper for the mutex separately, if we were to have it and the map itself as two separate fields of `KernelNameBasedCacheT`.

Reply:
Ah, you are right. Then let's create a separate structure instead of `std::pair<...>` and pass it to the `KernelProgramCache`.

I also don't like the idea:
> passing mutex and map hints separately). We would also need to handle creation/destruction in the wrapper for the mutex separately